pkg/services/object/put/distributed.go

```diff
@@ -211,95 +145,247 @@ func (t *distributedTarget) sendObject(node nodeDesc) error {
 	return nil
 }
 
-func (t *distributedTarget) iteratePlacement(f func(nodeDesc) error) (oid.ID, error) {
-	id, _ := t.obj.ID()
-	var resErr atomic.Value
+type errNotEnoughNodes struct {
+	listIndex int
+	required  uint
+	left      uint
+}
 
-loop:
-	for {
-		addrs := t.traverser.Next()
-		if len(addrs) == 0 {
-			break
-		}
+func (x errNotEnoughNodes) Error() string {
+	return fmt.Sprintf("number of replicas cannot be met for list #%d: %d required, %d nodes remaining",
+		x.listIndex, x.required, x.left)
+}
 
-		wg := new(sync.WaitGroup)
+type placementIterator struct {
+	log        *zap.Logger
+	neoFSNet   NeoFSNetwork
+	localPool  util.WorkerPool
+	remotePool util.WorkerPool
+	/* request-dependent */
+	containerNodes ContainerNodes
+	localOnly      bool
+	localNodePos   [2]int // in containerNodeSets. Undefined localOnly is false
+	// when non-zero, this setting simplifies the object's storage policy
+	// requirements to a fixed number of object replicas to be retained
+	linearReplNum uint
+	// whether to perform additional best-effort of sending the object replica to
+	// all reserve nodes of the container
+	broadcast bool
+}
 
-		for i := range addrs {
-			if t.traversalState.processed(addrs[i]) {
-				// it can happen only during additional container broadcast
+func (x placementIterator) iterateNodesForObject(obj oid.ID, f func(nodeDesc) error) error {
```

Comment:
can you, please, explain why this refactor was required (in commit message too)? does it solve something? the previous version was split into subfunctions and nesting was not that deep and the whole `iteratePlacement` was shorter. overall, i would say the previous approaches were closer to the best practices

Reply:
done